awx/main/scheduler/task_manager.py

```diff
@@ -481,16 +428,166 @@ def generate_dependencies(self, undeped_tasks):
 
         return created_dependencies
 
+    def process_tasks(self):
+        deps = self.generate_dependencies(self.all_tasks)
+        self.generate_dependencies(deps)
+        self.subsystem_metrics.inc(f"{self.prefix}_pending_processed", len(self.all_tasks) + len(deps))
+
+    @timeit
```

Comment:
I'm still really bothered that we are putting timings on `_schedule` and not `schedule`, because this misses timings of on_commit actions, which we know are a constant point of failure. Sure, this could be handled by filing a followup issue, but I really don't want it to get lost or slip into the next release, because it undermines the very point of the task manager metrics. We can still hit the _dispatcher_ timeout limits, with no forewarning by the metrics. Until we fix this, the metrics are not really to be believed.

just the other day, I was able to see this by a new mechanism:

```
tools_awx_1 | 2022-08-08 17:40:42,465 ERROR    [252b9dcc] awx.main.dispatch Worker failed to run task awx.main.scheduler.tasks.run_task_manager(*[], **{}
tools_awx_1 | Traceback (most recent call last):
tools_awx_1 |   File "/awx_devel/awx/main/dispatch/worker/task.py", line 102, in perform_work
tools_awx_1 |     result = self.run_callable(body)
tools_awx_1 |   File "/awx_devel/awx/main/dispatch/worker/task.py", line 77, in run_callable
tools_awx_1 |     return _call(*args, **kwargs)
tools_awx_1 |   File "/awx_devel/awx/main/scheduler/tasks.py", line 15, in run_task_manager
tools_awx_1 |     TaskManager().schedule()
tools_awx_1 |   File "/awx_devel/awx/main/scheduler/task_manager.py", line 657, in schedule
tools_awx_1 |     start_time = time.time()
tools_awx_1 |   File "/var/lib/awx/venv/awx/lib64/python3.9/site-packages/django/db/transaction.py", line 290, in __exit__
tools_awx_1 |     connection.set_autocommit(True)
tools_awx_1 |   File "/var/lib/awx/venv/awx/lib64/python3.9/site-packages/django/db/backends/base/base.py", line 420, in set_autocommit
tools_awx_1 |     self.run_and_clear_commit_hooks()
tools_awx_1 |   File "/var/lib/awx/venv/awx/lib64/python3.9/site-packages/django/db/backends/base/base.py", line 653, in run_and_clear_commit_hooks
tools_awx_1 |     func()
tools_awx_1 |   File "/awx_devel/awx/main/scheduler/task_manager.py", line 294, in post_commit
tools_awx_1 |     task_cls.apply_async(
tools_awx_1 |   File "/awx_devel/awx/main/dispatch/publish.py", line 87, in apply_async
tools_awx_1 |     with pg_bus_conn() as conn:
tools_awx_1 |   File "/usr/lib64/python3.9/contextlib.py", line 119, in __enter__
tools_awx_1 |     return next(self.gen)
tools_awx_1 |   File "/awx_devel/awx/main/dispatch/__init__.py", line 50, in pg_bus_conn
tools_awx_1 |     conn = psycopg2.connect(dbname=conf['NAME'], host=conf['HOST'], user=conf['USER'], password=conf['PASSWORD'], port=conf['PORT'], **conf.get("OPTIONS", {}))
tools_awx_1 |   File "/var/lib/awx/venv/awx/lib64/python3.9/site-packages/psycopg2/__init__.py", line 126, in connect
tools_awx_1 |     conn = _connect(dsn, connection_factory=connection_factory, **kwasync)
tools_awx_1 | psycopg2.OperationalError: FATAL:  sorry, too many clients already
```

This shows that the `on_commit` logic is even _more_ vulnerable than thought, because it requires establishing a new database connection. When approaching connection limits, this guarantees we leave jobs stuck in waiting. That may or may not be exasperated failing to close unused connections, not sure yet.